### DISCUSSION TITLE
[v5] Cannot resolve "Adonis/Core/Drive" namespace from the IoC Container

### DISCUSSION THREAD

**Discussion by wemersonrv** [adonisjs/core#3116] [Help]
Hello,

Cannot use Drive. It gives me error *`E_IOC_LOOKUP_FAILED: Cannot resolve "Adonis/Core/Drive" namespace from the IoC Container`*

What i am doing wrong?

```js
import Drive from '@ioc:Adonis/Core/Drive'

// ...

await Drive.put(myTmpPath, JSON.stringify({ foo: 'bar' }))

```

```bash
UPDATE: app/Services/WaaaService/index.ts
[ info ]  re-starting http server...

IocLookupException: E_IOC_LOOKUP_FAILED: Cannot resolve "Adonis/Core/Drive" namespace from the IoC Container

at /Users/wemersonrv/desenvolvimento/projetos/wemersonrv/waaa/app/Services/WaaaService/index.ts(anonymous):9
4   import { compose } from '@ioc:Adonis/Core/Helpers'
5   
6   import { ClientMixin } from './Mixins/ClientMixin'
7   import { ClientWithEventsMixin } from './Mixins/ClientWithEventsMixin'
8   
9   import Drive from '@ioc:Adonis/Core/Drive'
10  import { WaaaPaths } from 'Config/waaa'
11  
12  export default class WaaaService
13    extends compose(ClientMixin, ClientWithEventsMixin)
14    implements WaaaServiceContract

1 Function.lookupFailed
  /Users/wemersonrv/desenvolvimento/projetos/wemersonrv/waaa/node_modules/@adonisjs/fold/build/src/Exceptions/IocLookupException.js:18

2 Ioc.lookupOrFail
  /Users/wemersonrv/desenvolvimento/projetos/wemersonrv/waaa/node_modules/@adonisjs/fold/build/src/Ioc/index.js:254

3 Ioc.use
  /Users/wemersonrv/desenvolvimento/projetos/wemersonrv/waaa/node_modules/@adonisjs/fold/build/src/Ioc/index.js:308

4 Module._compile
  /Users/wemersonrv/desenvolvimento/projetos/wemersonrv/waaa/node_modules/pirates/lib/index.js:99

5 Object.newLoader [as .ts]
  /Users/wemersonrv/desenvolvimento/projetos/wemersonrv/waaa/node_modules/pirates/lib/index.js:104
```

**Comment by thetutlage** [comment]
- What's the version of `@adonisjs/core`?
- And where do you use `Services/WaaaService/index` service?

**Reply by wemersonrv** [reply]
> * What's the version of `@adonisjs/core`?

**v5.1.11**


> * And where do you use `Services/WaaaService/index` service?

This is the main file of my service. It\s added on IoC by a service provider. The first call to it is on the provider at `boot` method. See here: [#3095](https://github.com/adonisjs/core/discussions/3095#discussioncomment-1255106)

**Reply by thetutlage** [reply]
If you follow the release notes https://docs.adonisjs.com/releases/august-2021-release you should be using `5.3.x` and also please share the code for the provider. Coz, if you are importing your service as a top level import, then it will not work, coz the container is not ready at that point in time

**Reply by wemersonrv** [reply]
> If you follow the release notes https://docs.adonisjs.com/releases/august-2021-release you should be using `5.3.x` and also please share the code for the provider. Coz, if you are importing your service as a top level import, then it will not work, coz the container is not ready at that point in time

Updated to `v5.3.2`, and the problem still happens... below my provider

### providers/WaaaProvider.ts
```ts
import { ApplicationContract } from '@ioc:Adonis/Core/Application'
import WaaaService from 'App/Services/WaaaService'

export default class WaaaProvider {
  constructor(protected app: ApplicationContract) {}

  public register() {
    this.app.container.singleton('@ioc:App/Services/WaaaService', function () {
      return new WaaaService()
    })
  }

  public async boot() {
    this.app.container.use('@ioc:App/Services/WaaaService')
  }

  public async ready() {}

  public async shutdown() {}
}
```

### app/Services/WaaaService.ts
```ts
import { WaaaServiceContract } from '@ioc:App/Services/WaaaService'

import Drive from '@ioc:Adonis/Core/Drive'
import { WaaaPaths } from 'Config/waaa'

export default class WaaaService implements WaaaServiceContract
{
  constructor() {
    this._testDrive()
  }

  protected async _testDrive() {
    const path = WaaaPaths.sessionPath
    await Drive.put(path, JSON.stringify({ foo: 'bar' }))
  }
}
```

**Reply by thetutlage** [reply]
As mentioned earlier, you cannot have top level imports in your service provider that relies on container bindings. 

Also, I am not sure why you are even binding your service to the container vs using it directly like any other regular class

**Reply by wemersonrv** [reply]
> As mentioned earlier, you cannot have top level imports in your service provider that relies on container bindings.
> 
> Also, I am not sure why you are even binding your service to the container vs using it directly like any other regular class

My project relies on this lib for Whatsapp: [whatsapp-web.js](https://github.com/pedroslopez/whatsapp-web.js). it has some realtime events on backend that i need to list when app starts. And need to use same object when receive HTTP requests, because if create instances for it each request will duplicate event listeners, and another unexpected issues... so make sense to me to add it as a singleton to container, and reuse it.

I already have it running in adonis v4 for a year, and works fine as a singleton on IoC. 

The problem i had now is that i just need to read a json file with some settings when this class is started. Just that... If i cannot use it with Drive, what other option to access filesystem i have? In adonis v4 i used Helpers Provider to achieve that with `Helpers.promisify(require('fs'))`
